climada/engine/uncertainty/unc_impact.py

```diff
@@ -0,0 +1,309 @@
+            chunksize = min(self.n_runs // pool.ncpus, 100)
+            imp_metrics = pool.map(self._map_impact_calc,
+                                           self.samples.iterrows(),
+                                           chunsize = chunksize)
+
+        else:
+            imp_metrics = map(self._map_impact_calc, self.sample.iterrows())
+        
+        #Perform the actual computation
+        with log_level(level='ERROR', name_prefix='climada'):
+            [aai_agg_list, freq_curve_list,
+             eai_exp_list, at_event_list,
+             tot_value_list] = list(zip(*imp_metrics))
+            
+
+        # Assign computed impact distribution data to self
+        self.metrics['aai_agg']  = pd.DataFrame(aai_agg_list,
+                                                columns = ['aai_agg'])
+
+        self.metrics['freq_curve'] = pd.DataFrame(freq_curve_list,
+                                    columns=['rp' + str(n) for n in rp])
+        self.metrics['eai_exp'] =  pd.DataFrame(eai_exp_list)
+        self.metrics['at_event'] = pd.DataFrame(at_event_list)
+        self.metrics['tot_value'] = pd.DataFrame(tot_value_list,
+                                                 columns = ['tot_value'])
+        self.check()
+        
+        return None
+
+    def _map_impact_calc(self, sample_iterrows):
+        """
+        Map to compute impact for all parameter samples in parrallel
+
+        Parameters
+        ----------
+        sample_iterrows : pd.DataFrame.iterrows()
+            Generator of the parameter samples
+
+        Returns
+        -------
+         : list
+            impact metrics list for all samples containing aai_agg, rp_curve,
+            eai_exp (np.array([]) if self.calc_eai_exp=False) and at_event
+            (np.array([]) if self.calc_at_event=False).
+
+        """
+
+        # [1] only the rows of the dataframe passed by pd.DataFrame.iterrows()
+        exp_samples = sample_iterrows[1][self.unc_vars['exp'].labels].to_dict()
+        haz_samples = sample_iterrows[1][self.unc_vars['haz'].labels].to_dict()
+        impf_samples = sample_iterrows[1][self.unc_vars['impf'].labels].to_dict()
+
+        exp = self.unc_vars['exp'].evaluate(exp_samples)
+        haz = self.unc_vars['haz'].evaluate(haz_samples)
+        impf = self.unc_vars['impf'].evaluate(impf_samples)
+
+        imp = Impact()
+
+        imp.calc(exposures=exp, impact_funcs=impf, hazard=haz)
+    
+        # Extract from climada.impact the chosen metrics
+        freq_curve = imp.calc_freq_curve(self.rp).impact
+
+        if self.calc_eai_exp:
+            eai_exp = imp.eai_exp
+        else:
+            eai_exp = np.array([])
+
+        if self.calc_at_event:
+            at_event= imp.at_event
+        else:
+            at_event = np.array([])
+
+        return [imp.aai_agg, freq_curve, eai_exp, at_event, imp.tot_value]
+
+
+    def plot_rp_distribution(self):
+        """
+        Plot the distribution of return period values.
```

Comment:
Perhaps one more sentence at what stage this plotting method works (i.e. what needs to be computed beforehand)

Reply:
I am not sure whether this makes too much sense now. For example, in the `Impact` class, there is no mention that you need to run `.calc` before you can plot... Why should one add such points in particular here? There is already a check with an error if the distribution data is not present

```
        if not self.metrics:
            raise ValueError("No uncertainty data present for these metrics. "+
                    "Please run an uncertainty analysis first.")
```

Reply:
yes, fair